plutus-core/plutus-core/src/PlutusCore/TypeCheck/Internal.hs

```diff
@@ -523,7 +523,9 @@ inferTypeM t@(Constr ann resTy i args) = do
 
     -- We don't know exactly what to expect, we only know what the i-th sum should look like, so we
     -- assert that we should have some types in the sum up to there, and then the known product type.
-    let prodPrefix  = map (\j -> "prod_" <> Text.pack (show j)) [0 .. i - 1]
+    let -- 'toInteger' is necessary, because @i@ is a @Word64@ and therefore @i - 1@ would be
+        -- @maxBound :: Word64@ for @i = 0@ if we didn't have 'toInteger'.
+        prodPrefix  = map (\j -> "prod_" <> Text.pack (show j)) [0 .. toInteger i - 1]
```

Comment:
Oh, that's pretty impressive.  I was expecting some subtle error in the typechecking algorithm, but it's just the _error message_.  That's an easy mistake to make: I'd forgotten that Word64 wraps round at the bounds.